versions/1.1/SPEC.md

```diff
@@ -0,0 +1,5967 @@
+
+Some functions accept arguments of multiple different types, denoted as a list of types separated by `|`.
+
+Some functions are polymorphic, which means that they are actually multiple functions with the same name but different signatures. Such functions are defined with generic types (e.g. `X`, `Y`) instead of concrete types (e.g. `File` or `String`), and the bounds of each type parameter is specified in the function description.
+
+Functions that are new in this version of the specification are denoted by ✨, and deprecated functions are denoted by 🗑.
+
+## Int floor(Float), Int ceil(Float) and Int round(Float)
+
+These functions convert a `Float` to an `Int` using different rounding methods:
+
+- `floor`: Rounds **down** to the next lower integer.
+- `ceil`: Rounds **up** to the next higher integer.
+- `round`: Rounds to the nearest integer based on standard rounding rules ("round half up").
+
+**Parameters**:
+
+1. `Float`: the number to round.
+
+**Returns**: An integer.
+
+**Example**
+
+```wdl
+# all these expressions evaluate to true
+Boolean all_true = [
+  floor(1.0) == 1,
+  floor(1.9) == 1,
+  ceil(2.0) == 2,
+  ceil(2.1) == 3,
+  round(1.49) == 1,
+  round(1.50) == 2
+]
+```
+
+## ✨ Int min(Int, Int), Float min(Float, Float), Float min(Int, Float), Float min(Float, Int)
+
+Returns the smaller of two values. If both values are `Int`s, the return value is an `Int`, otherwise it is a `Float`.
+
+**Parameters**:
+
+1. `Int|Float`: the first number to compare.
+2. `Int|Float`: the second number to compare.
+
+**Returns**: The smaller of the two arguments.
+
+**Example**
+
+```wdl
+workflow min_test {
+  input {
+    Int value1
+    Float value2
+  }
+  output {
+    # these two expressions are equivalent
+    Float min1 = if value1 < value2 then value1 else value2
+    Float min2 = min(value1, value2)
+  }
+}
+``` 
+
+## ✨ Int max(Int, Int), Float max(Float, Float), Float max(Int, Float), Float max(Float, Int)
+
+Returns the larger of two values. If both values are `Int`s, the return value is an `Int`, otherwise it is a `Float`.
+
+**Parameters**:
+
+1. `Int|Float`: the first number to compare.
+2. `Int|Float`: the second number to compare.
+
+**Returns**: The larger of the two arguments.
+
+**Example**
+
+```wdl
+workflow max_test {
+  input {
+    Int value1
+    Float value2
+  }
+  output {
+    # these two expressions are equivalent
+    Float max1 = if value1 > value2 then value1 else value2
+    Float max2 = max(value1, value2)
+  }
+}
+``` 
+
+## String sub(String, String, String)
+
+Given 3 String parameters `input`, `pattern`, `replace`, this function replaces all non-overlapping occurrences of `pattern` in `input` by `replace`. `pattern` is a [regular expression](https://en.wikipedia.org/wiki/Regular_expression) that will be evaluated as a [POSIX Extended Regular Expression (ERE)](https://en.wikipedia.org/wiki/Regular_expression#POSIX_basic_and_extended).
```

Comment:
How should the regex string be interpreted regarding escape characters?

I.e. in WDL a string with a backslash is `\\`. Where the former backslash escapes the second. So now for regex there are two possibilities.

1. The regex is written as a WDL string. In POSIX ERE a backslash needs to be escaped, so it is `\\`. When the regex is written as a WDL string it needs to be written as `\\\\`. The engine then converts the string properly to `\\`. The regex interpreter then knows it should match `\`.
2. The regex is written as a raw string. Meaning `\\` will translate to a regex `\\`, which matches `\`. 

I prefer option 2. But there is some ambiguity about this. See the issue @mlin posted on biowdl/tasks: https://github.com/biowdl/tasks/issues/270.

Reply:
Speaking from the perspective of the parser, option 1 is much cleaner. Otherwise we need to put in a special case for every function that accepts a regexp OR have a special raw string prefix like in python (e.g. r""). For now I say we go with option 1 and consider the change for option 2 in the 2.0 spec.

Reply:
The point is that this changed between two releases of cromwell at some point. Whichever option is chosen it should be unambiguous from the spec. 
I can understand the parser perspective, but for the user writing `\\\\` to mean `\` is quite a nightmare. 

> The Workflow Description Language (WDL) is a way to specify data processing workflows with a **human-readable and -writeable syntax**. WDL makes it straightforward to define analysis tasks, chain them together in workflows, and parallelize their execution. The language makes common patterns **simple to express**, while also admitting uncommon or complicated behavior; and strives to achieve portability not only across execution platforms, but also different types of users. Whether one is an **analyst, a programmer, an operator of a production system, or any other sort of user**, WDL should be accessible and understandable. 

emphasis mine. I think it would be more fitting with the WDL philosophy to go for option 2. People can look up how regexes work, and their copy-pasted regex examples will work as intended without needing to worry about escaping their regex escape characters. 
Currently cromwell does support the raw string option (at least it seems to work that way). So it would be a shame to lose that. But that is just my 2 cents.

Reply:
Yes, but Cromwell != WDL.

I agree that it should be unambiguous, but I don't think it's appropriate to add it in this version of the spec.

Reply:
Option 1 is far better than leaving it ambiguous. And if it is easier to implement it will mean we will get v1.1 implementation sooner which is also desirable.
Adding raw strings later is a very good idea. A raw string construct makes it easier for the parser as well as the user, while being explicit as well.

Reply:
Great - do you want to make a PR for raw strings?

Reply:
Yup, I will put it on my todo list :slightly_smiling_face: .